### PR TITLE
statedb: Simplify integration with Hive

### PR DESCRIPTION
**Pull request by joamaki**
Instead of relying on the group values use an explicit registration for the tables. This allows more flexibility for how to create the tables and how to for example scope them to a module or even to keep RWTable[T] hidden inside an object without having the complicated `NewProtectedTableCell` and so on helpers which obfuscate what is going on.

```
// A constructor function for a table. Used by the real implementation and integration tests of components
// that depend on this table can use this to create and populate it.
func NewExampleTable() statedb.RWTable[*Example] {
  return statedb.NewTable[*Example]("test", ExampleIDIndex, ExampleNameIndex)
}

func example() {
  hive.New(
    statedb.Cell,
    cell.Module("example", "Example",
      cell.ProvidePrivate(NewExampleTable), // Provides RWTable[*Example] just to this module.

      // 1st way to register a table.
      cell.Invoke(statedb.RegisterTable[*Example]),

      // 2nd longer way to register a table
      cell.Invoke(func(db *statedb.DB, table statedb.RWTable[*Example]) error {
        return db.RegisterTable(table)
      }),       

    // Provide the read-only "Table[*Example]" API to the whole hive. Here or from a start
    // hook we can for example populate it before anything reads it. We get the guarantee
    // that the start hook is called before any readers if we append the hook from this constructor.
    cell.Provide(func(table statedb.RWTable[*Example]) statedb.Table[*Example] {
      return table
    }),

    // Shorter way to provide Table[*Example] using ToTable.
    // This method referenced this way has type "func(table RWTable[*Example]) Table[*Example]"
    cell.Provide(statedb.RWTable[*Example].ToTable),
  )
}

### CODE REVIEW

File: pkg/statedb/table.go [hunk 1]
```diff
@@ -13,11 +13,14 @@ import (
 	"github.com/cilium/cilium/pkg/statedb/index"
 )
 
+// NewTable creates a new table with given name and indexes.
+// Panics if indexes are malformed which allows this to be used at
```
Comment (bimmlerd) [inline pkg/statedb/table.go]: Is this worth the risk of a runtime panic?

As far as I can tell, it's now possible to create and register tables at an arbitrary point at runtime, and there might even be a legitimate use-case for it in the future. If that's so, then I'd much rather have to deal with an error than have a potential panic, even if it leads to slightly uglier code in the tests.

**Review by bimmlerd — CHANGES_REQUESTED** [review]
I don't like the potential `panic`s in `NewTable`.

**Review by dylandreimerink — APPROVED** [review]
LGTM

File: pkg/statedb/table.go (hunk 1 above)
Comment (dylandreimerink) [reply pkg/statedb/table.go]: That is a good point, this does invite table creation after the constructor/start hooks where panics are less ideal. Even though its not how we intend it, that is going to be inevitable.

**Review by bimmlerd — APPROVED** [review]
Thanks!

File: pkg/statedb/table.go (hunk 1 above)
Comment (joamaki) [reply pkg/statedb/table.go]: Agreed, changed it to return error and added `MustNewTable` for the few cases where we prefer that in the tests.

**Comment by bimmlerd** [comment]
CI triage:
- Conformance Ginkgo (ci-ginkgo): hit a deadlock https://github.com/cilium/cilium/issues/29078 (rerunning after fixing the conflicts, it's an unlikely one)
- Conformance AWS-CNI (ci-awscni): seems to have hit https://github.com/cilium/cilium/issues/28625 (rerunning after fixing conflicts)

**Comment by bimmlerd** [comment]
CI triage:
- ci-ginkgo failed on `E2E Test (1.28, f09-datapath-misc-2)`, but one of the cluster nodes just didn't come up, cilium agent never go to run on there, hence unrelated and rerunning.
- GKE failed with `connectivity test failed: unmarshaling cilium runtime config json: unexpected end of JSON input` for one of three - sounds like an infra issue. Not required.
- AKS failed a connectivity test `[pod-to-pod-encryption/pod-to-pod-encryption/curl-ipv4: cilium-test/client-6b4b857d98-ktkp4 (10.0.1.192) -> cilium-test/echo-other-node-6f988f8697-mhm6k (10.0.0.64:8080)]` with `❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://10.0.0.64:8080/" failed: error dialing backend: EOF`, but is also not required.

**Pull request merged**